File: Array/2501. Longest Square Streak in an Array.cpp
/*
2501. Longest Square Streak in an Array
*/

// TC:O(nlgn) SC:O(n)

class Solution {
public:
    int longestSquareStreak(vector<int>& nums) {
        sort(nums.begin(), nums.end());

        int n = nums.size();
        unordered_map<int,int> map;
        map[nums[0]] = 1;
        int res = -1;
        for(int i = 1; i < n; i++) {
            int cur = sqrt(nums[i]);
            if(cur * cur == nums[i]) {
                if(map.find(cur) != map.end()) {
                    map[nums[i]] = map[cur] + 1;
                    if(map[nums[i]] >= 2) res = max(res, map[nums[i]]);
                }
                else map[nums[i]] = 1;
            }
            else map[nums[i]] = 1;
        }

        return res;
    }
};

File: Dynamic_Programming/2501. Longest Square Streak in an Array.py
# Math: TC:O(nlglgU) SC:O(n)
class Solution:
    def longestSquareStreak(self, nums: List[int]) -> int:
        s = set(nums)
        ans = -1
        for x in s:
            cnt = 0
            while x in s:
                cnt += 1
                x *= x
            if cnt >= 2 and cnt > ans:
                ans = cnt

        return ans

# DP TC:O(n) SC:O(n)
class Solution:
    def longestSquareStreak(self, nums: List[int]) -> int:
        s = set(nums)
        @cache
        def dfs(x:int) ->int:
            if x not in s:
                return 0
            return 1 + dfs(x * x)
        ans = max(map(dfs, s))
        return ans if ans > 1 else -1
